### PR TITLE
Use MutationObserver to listen for DOM changes.

### LINKED ISSUE #601: Very slow performance on page with Facebook Container extension enabled
- Facebook Container Version: 2.0.3
- Operating System + Version: macOS 10.15.3
- Firefox Version: 75.0a1 (2020-02-18) (64-bit)
- Other installed Add-ons + Version + Enabled/Disabled-Status: No other extensions enabled with this profile

When using the accordions on the page https://developer.mypurecloud.com/api/rest/v2/conversations/index.html, the web page freezes with Facebook Container enabled. With the extension disabled, the page is relatively smooth.

### Actual behavior
Web page is very slow. Profile: https://perfht.ml/38HAxnk

### Expected behavior
Expected web page to be smooth, which it is with the Facebook Container extension disabled.

### Steps to reproduce
1. Enable Facebook Container extension
2. Visit https://developer.mypurecloud.com/api/rest/v2/conversations/index.html
3. Open one of the accordion items (example in the profile is the `GET /api/v2/conversations/messages/{conversationId}/messages/{messageId}` accordion item)
4. Notice web page freezes for a few seconds before the item opens.
5. Repeat with Facebook Container disabled, and notice performance is smooth.

### PR DESCRIPTION
**Pull request by mrwensveen**
- Only run pattern detection on changed nodes.
- Start the content script at document_start, otherwise we miss the window.load event (not 100% sure about this one)
- Fix "Badging" feature test.

Please note the part with "XXX: ..". I needed to add a timeout for Selenium to work, but when manually testing the same page (badges.html) it worked fine without the timeout. It seems that `browser.runtime.sendMessage` is too early in some scenarios.

fixes #601

### CODE REVIEW

**Comment by mrwensveen** [comment]
Hi. I merged the latest from the upstream main branch into my own branch (and PR) so that there aren't any merge conflicts.

I do see a failing test, which I would like to look at. How do I run these tests locally? Or is there another way to look at the test result and fix this problem?

**Comment by maxxcrawford** [comment]
@codemist Can you incorporate this PR into your iframe work? I'd love to merge this in. Let's sync and see if we should merge this in as is first, and you solve conflicts from your side.

**Comment by maxxcrawford** [comment]
> Hi @maxxcrawford,
> 
> I merged the latest from the upstream main branch into my own branch (and PR) so that there aren't any merge conflicts.
> 
> ~I do see a failing test, which I would like to look at. How do I run these tests locally? Or is there another way to look at the test result and fix this problem?~ It looks like I already fixed that problem in the 2nd commit, so there shouldn't be anything in the way to merge this.
> 
> Cheers, Matthijs

@mrwensveen Thanks for the work on this. We will incorporate this in our next release!

**Review by codemist — COMMENTED** [review]
Tested this and it worked except for missing FBC icon on email fields. On L702,  `patternDetection(EMAIL_PATTERN_DETECTION_SELECTORS, "email")` is missing an argument, should be  `patternDetection(EMAIL_PATTERN_DETECTION_SELECTORS, "email", target)`.

**Comment by mrwensveen** [comment]
Thanks @codemist! Should be fixed now.

File: src/content_script.js [hunk 1]
```diff
@@ -74,6 +74,12 @@ const PASSIVE_SHARE_PATTERN_DETECTION_SELECTORS = [
   "[href*='facebook.com/sharer']", // Legacy Share dialog
 ];
 
+// Attributes distilled from selectors above. Update when necessary.
+const OBSERVER_ATTRIBUTES = [
+  "action", "aria-label", "class",
+  "data-action", "data-bfa-network", "data-destination", "data-login-with-facebook",
+  "data-oauthserver", "data-partner", "data-tag", "data-test-id", "data-tracking",
+  "href", "id", "title"];
```
Comment (maxxcrawford) [inline src/content_script.js:82]: praise: This is so much cleaner to watch these attrs, rather the whole, dang DOM.

File: src/content_script.js [hunk 2]
```diff
@@ -833,15 +845,46 @@ async function CheckIfURLShouldBeBlocked() {
...
+    });
   }
 
 }
 
 // Cross-browser implementation of element.addEventListener()
 function addPassiveWindowOnloadListener() {
   window.addEventListener("load", function() {
-    CheckIfURLShouldBeBlocked();
+    // XXX: Work around slow test startup.
```
Comment (maxxcrawford) [inline src/content_script.js:879]: suggestion: (non-blocking) Can we change this from XXX to TODO/FIXME?

**Review by maxxcrawford — APPROVED** [review]
@mrwensveen Thank you for this PR! This looks good to ship. I had one nit comment, but we can resolve it in a future PR. 

_Commented using [conventional comments](conventionalcomments.org) language_

**Pull request merged**